app/src/main/java/org/mozilla/focus/web/WebViewProvider.kt

```diff
@@ -0,0 +1,83 @@
+package org.mozilla.focus.web
+
+import android.content.Context
+import android.preference.PreferenceManager
+import android.util.AttributeSet
+import android.view.View
+import android.webkit.WebSettings
+import org.mozilla.focus.web.Renderers.WEBVIEW
+import org.mozilla.focus.web.Renderers.GECKO
+import org.mozilla.focus.webview.SystemWebView
+
+const val ENGINE_PREF_STRING_KEY = "engine_choice"
+
+object WebViewProvider {
+    var engineName: String = Config.renderer.name
+    var engine : IWebViewProvider? = null
+
+    fun writeEnginePref(context: Context, newEngineName: String = engineName) {
+        requireNotNull(context)
+        val prefs = PreferenceManager.getDefaultSharedPreferences(context)
+        prefs.edit().putString(ENGINE_PREF_STRING_KEY, newEngineName).apply()
+    }
+
+    fun readEnginePref(context: Context) {
+        requireNotNull(context)
+        val prefs = PreferenceManager.getDefaultSharedPreferences(context)
+        engineName = prefs.getString(ENGINE_PREF_STRING_KEY, Config.renderer.name)
+        useEngine(engineName)
+    }
+
+    private fun useEngine(name: String) {
+        when (name) {
+            WEBVIEW.name -> engine = ClassicWebViewProvider()
+            GECKO.name -> engine = GeckoWebViewProvider()
+        }
+    }
+
+    fun preload(context: Context) {
+        requireNotNull(context)
+        requireNotNull(engine)
+        engine!!.preload(context)
```

Comment:
Also here I do not understand the usage of `requireNotNull` and `!!`. Wouldn't just using `!!` have the same result? Or using the return value of `requireNotNull`?

Reply:
Since engine is nullable, I either need to use !!. or ?. to make the call safe. It does seem redundant in this case. I can remove the precondition.